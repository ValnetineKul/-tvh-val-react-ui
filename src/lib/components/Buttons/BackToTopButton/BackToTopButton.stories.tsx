import type { ComponentProps } from 'react';
import React from 'react';
import type { StoryTemplate } from '../../../story-utils';
import { createMeta } from '../../../story-utils';
import BackToTopButton from './BackToTopButton';

type Props = ComponentProps<typeof BackToTopButton>;

const Template = ((args: Props) => (
  <div style={{ padding: 16, height: window.screen.height * 4 }}>
    <p>The button appears when users scroll 2 pages down and move the scroll bar up or scroll up on the page.</p>
    <BackToTopButton {...args} />
  </div>
)) as StoryTemplate<Props>;

export const Primitive = Template.bind({});

Primitive.storyName = 'BackToTopButton';

export default createMeta({
  component: BackToTopButton,
  title: 'Components/Buttons/BackToTopButton',
  argTypes: {
    icon: {
      control: {
        disable: true,
      },
      table: {
        type: { summary: 'Icon with FontAwesome definition' },
      },
    },
  },
  parameters: {
    componentSubtitle: '',
  },
});
