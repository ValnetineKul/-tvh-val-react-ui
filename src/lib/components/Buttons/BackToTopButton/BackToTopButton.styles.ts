import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles<void, 'backToTop'>({ name: 'BackToTopButton' })((theme, _params, classes) => ({
  root: {
    '&:disabled': {
      backgroundColor: theme.color['bg/disabled'],
    },
    [`.MuiButtonBase-root + &.${classes.backToTop}`]: {
      bottom: `calc(${theme.layoutSpacing['spacing/300']} + 40px + ${theme.layoutSpacing['spacing/300']})`,
      [theme.breakpoints.up('sm')]: {
        bottom: `calc(${theme.layoutSpacing['spacing/450']} + 56px + ${theme.layoutSpacing['spacing/400']})`,
      },
    },
  },
  backToTop: {},
}));

export default useStyles;
