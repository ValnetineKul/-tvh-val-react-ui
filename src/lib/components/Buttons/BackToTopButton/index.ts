import BackToTopButton from './BackToTopButton';

export type { BackToTopButtonProps } from './BackToTopButton';
export default BackToTopButton;
