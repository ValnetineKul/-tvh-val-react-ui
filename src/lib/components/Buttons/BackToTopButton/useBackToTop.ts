import { useState, useEffect, useCallback } from 'react';

const useBackToTop = () => {
  const [scrollPosition, setScrollPosition] = useState(0);
  const [shouldShowBackToTop, setShowBackToTop] = useState(false);
  const [shouldUpScroll, setShouldUpScroll] = useState(false);

  const handleScroll = useCallback(() => {
    if (scrollPosition > window.scrollY) {
      if (!shouldUpScroll) {
        setShouldUpScroll(true);
      }
    } else if (shouldUpScroll) {
      setShouldUpScroll(false);
    }
    setScrollPosition(window.scrollY);
  }, [scrollPosition, shouldUpScroll]);

  useEffect(() => {
    window.addEventListener('scroll', handleScroll, { passive: true });
    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, [handleScroll, scrollPosition, shouldUpScroll]);

  useEffect(() => {
    const border = window.screen.height * 2;

    if (scrollPosition > border && shouldUpScroll) {
      setShowBackToTop(true);
    } else if (scrollPosition === 0 || (!shouldUpScroll && shouldShowBackToTop)) {
      setShowBackToTop(false);
    }
  }, [scrollPosition, shouldShowBackToTop, shouldUpScroll]);

  return shouldShowBackToTop;
};

export default useBackToTop;
