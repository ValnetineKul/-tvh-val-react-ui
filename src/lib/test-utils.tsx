import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
