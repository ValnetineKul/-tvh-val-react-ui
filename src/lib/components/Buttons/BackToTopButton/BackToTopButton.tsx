import type { FC } from 'react';
import React from 'react';
import FloatingActionButton from '../FloatingActionButton';
import { AngleUp } from '../../Icon/icons/functional';
import Icon from '../../Icon';
import useStyles from './BackToTopButton.styles';
import useBackToTop from './useBackToTop';

export interface BackToTopButtonProps extends React.AriaAttributes {
  disabled?: boolean;
  className?: string;
}

const BackToTopButton: FC<BackToTopButtonProps> = ({ disabled = false, className, ...props }) => {
  const { classes, cx } = useStyles();

  const shouldShowBackToTop = useBackToTop();
  if (!shouldShowBackToTop) {
    return null;
  }

  return (
    <FloatingActionButton
      disabled={disabled}
      className={cx(classes.root, classes.backToTop, className)}
      onClick={() => window.scrollTo({ top: 0, behavior: 'smooth' })}
      icon={<Icon icon={AngleUp} />}
      variant="secondary"
      {...props}
    />
  );
};

export default BackToTopButton;
