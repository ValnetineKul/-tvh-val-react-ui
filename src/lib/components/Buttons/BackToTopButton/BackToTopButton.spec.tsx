import React from 'react';
import useBackToTop from './useBackToTop';
import { render, userEvent } from '../../../test-utils';
import BackToTopButton from './BackToTopButton';

jest.mock('./useBackToTop');

describe('BackToTopButton', () => {
  it('Should render correctly', () => {
    (useBackToTop as jest.Mock).mockReturnValue(true);
    const { container } = render(<BackToTopButton />);
    expect(container).toMatchSnapshot();
  });

  it('Should get callback on click', () => {
    (useBackToTop as jest.Mock).mockReturnValue(true);
    window.scrollTo = jest.fn();
    const { getByRole } = render(<BackToTopButton />);
    const button = getByRole('button');
    userEvent.click(button);
    expect(window.scrollTo).toBeCalledTimes(1);
    expect(window.scrollTo).toHaveBeenCalledWith({ top: 0, behavior: 'smooth' });
  });

  it('Should not render if not needed', () => {
    (useBackToTop as jest.Mock).mockReturnValue(false);
    const { queryByRole } = render(<BackToTopButton />);
    const button = queryByRole('button');
    expect(button).not.toBeInTheDocument();
  });
});
