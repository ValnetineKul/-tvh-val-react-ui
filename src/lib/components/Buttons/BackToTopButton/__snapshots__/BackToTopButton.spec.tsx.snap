// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTopButton Should render correctly 1`] = `
<div>
  <button
    class="MuiButtonBase-root tss-3-BackToTopButton-backToTop-ref tss-1qpp0so-ButtonBase-root-ButtonBase-focusOut-FloatingActionButton-root-BackToTopButton-root-BackToTopButton-backToTop mui-10d1a0h-MuiButtonBase-root"
    tabindex="0"
    type="button"
  >
    <div
      class="tss-w1sjks-Icon-root-Icon-sizeMd"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
        color="inherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        angleUp.svg
      </svg>
    </div>
  </button>
</div>
`;
